tornado/tcpclient.py

```diff
@@ -116,24 +124,38 @@ def on_connect_done(self, addrs, af, addr, future):
                 self.on_timeout()
             return
         self.clear_timeout()
+        self.clear_connect_timeout()
         if self.future.done():
             # This is a late arrival; just drop it.
             stream.close()
         else:
             self.future.set_result((af, addr, stream))
 
-    def set_timout(self, timeout):
+    def set_timeout(self, timeout):
         self.timeout = self.io_loop.add_timeout(self.io_loop.time() + timeout,
                                                 self.on_timeout)
 
     def on_timeout(self):
         self.timeout = None
-        self.try_connect(iter(self.secondary_addrs))
+        if not self.future.done():
+            self.try_connect(iter(self.secondary_addrs))
 
     def clear_timeout(self):
         if self.timeout is not None:
             self.io_loop.remove_timeout(self.timeout)
 
+    def set_connect_timeout(self, connect_timeout):
+        self.connect_timeout = self.io_loop.add_timeout(
+            connect_timeout, self.on_connect_timeout)
+
+    def on_connect_timeout(self):
+        if not self.future.done():
+            self.future.set_exception(TimeoutError())
```

Comment:
One final piece of the puzzle: When we've timed out (or when we've succeeded after trying multiple times), we need to close the streams we've created. This means we need to change the interface between `TCPClient` and `_Connector`: `TCPClient._create_stream` should return a tuple `(stream, stream.connect(...)`, and the `_Connector` needs to keep a `set` of streams it has started.